### PR TITLE
Add support for proxy url whitelisting

### PR DESCRIPTION
**Pull request by Dalimil**
Users may only be interested in a few urls and with this patch can exclude everything from proxy and then only whitelist those few urls.

### CODE REVIEW

File: src/org/parosproxy/paros/core/proxy/ProxyServer.java [hunk 1]
```diff
@@ -399,13 +408,34 @@ public boolean excludeUrl(URI uri) {
                     if (log.isDebugEnabled()) {
                         log.debug("URL excluded: " + uriString + " Regex: " + p.pattern());
                     }
-                    
+
```
Comment (kingthorin) [inline src/org/parosproxy/paros/core/proxy/ProxyServer.java]: Formatting changes (if really necessary) should occur in a separate PR.

https://github.com/zaproxy/zaproxy/wiki/DevGuidelines

File: src/org/parosproxy/paros/model/Session.java [hunk 2]
```diff
@@ -52,7 +52,7 @@
 // ZAP: 2015/04/02 Issue 321: Support multiple databases and Issue 1582: Low memory option
 // ZAP: 2015/08/19 Change to use ZapXmlConfiguration instead of extending FileXML
 // ZAP: 2015/08/19 Issue 1789: Forced Browse/AJAX Spider messages not restored to Sites tab
-// ZAP: 2015/10/21 Issue 1576: Support data driven content
+// ZAP: 2015/10/21 Issue 1576: Support data driven content
```
Comment (kingthorin) [inline src/org/parosproxy/paros/model/Session.java]: What was changed here?

File: src/org/parosproxy/paros/view/View.java [hunk 3]
```diff
@@ -121,20 +121,7 @@
...
-import org.zaproxy.zap.view.ContextPanelFactory;
-import org.zaproxy.zap.view.ContextStructurePanel;
-import org.zaproxy.zap.view.ContextTechnologyPanel;
-import org.zaproxy.zap.view.SessionExcludeFromProxyPanel;
-import org.zaproxy.zap.view.SessionExcludeFromScanPanel;
-import org.zaproxy.zap.view.SessionExcludeFromSpiderPanel;
-import org.zaproxy.zap.view.SplashScreen;
-import org.zaproxy.zap.view.StatusUI;
-import org.zaproxy.zap.view.ZapMenuItem;
+import org.zaproxy.zap.view.*;
```
Comment (kingthorin) [inline src/org/parosproxy/paros/view/View.java]: Specific imports should be used.

File: src/org/zaproxy/zap/view/SessionIncludeInProxyPanel.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ * 
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ * 
+ * Copyright 2010 psiinon@gmail.com
```
Comment (kingthorin) [inline src/org/zaproxy/zap/view/SessionIncludeInProxyPanel.java]: This should be set as 2016 and "the ZAP Development Team" (check other recent PRs if you want to copy/paste)

File: src/org/zaproxy/zap/view/SessionIncludeInProxyPanel.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+package org.zaproxy.zap.view;
+
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.model.Model;
+import org.parosproxy.paros.model.Session;
+import org.parosproxy.paros.view.AbstractParamPanel;
+import org.parosproxy.paros.view.View;
+
+import javax.swing.*;
+import java.awt.*;
```
Comment (kingthorin) [inline src/org/zaproxy/zap/view/SessionIncludeInProxyPanel.java]: Should use specific imports.

File: src/org/zaproxy/zap/view/SessionIncludeInProxyPanel.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+	private MultipleRegexesOptionsPanel regexesPanel;
+
+    public SessionIncludeInProxyPanel() {
+        super();
+ 		initialize();
+	}
+
+	/**
+	 * This method initializes this
+	 */
```
Comment (kingthorin) [inline src/org/zaproxy/zap/view/SessionIncludeInProxyPanel.java]: Is this comment really needed?

File: src/org/zaproxy/zap/view/SessionIncludeInProxyPanel.java [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+	private JPanel getPanelSession() {
+		if (panelSession == null) {
+
+			panelSession = new JPanel();
+			panelSession.setLayout(new GridBagLayout());
+			panelSession.setName("Includeproxy");
+
+			GridBagConstraints gridBagConstraints2 = new GridBagConstraints();
+	        GridBagConstraints gridBagConstraints1 = new GridBagConstraints();
```
Comment (kingthorin) [inline src/org/zaproxy/zap/view/SessionIncludeInProxyPanel.java]: There seems to be some inconsistency in tabs/spaces here vs the previous few lines.

**Comment by kingthorin** [comment]
Thanks for tackling this. I've added a few quick hit comments, I haven't been able to do a detailed review.

Note any edited Paros file should be appropriately commented/documented: https://github.com/zaproxy/zaproxy/wiki/DevGuidelines

Also the PR should contain a single commit. Which can be accomplished through interactive rebase and fixup or squash.
https://www.kernel.org/pub/software/scm/git/docs/git-rebase.html#_interactive_mode
https://wiki.eclipse.org/EGit/User_Guide#Interactive_Rebase

**Comment by kingthorin** [comment]
Thanks for making this into a single commit. Could you address the feedback?

Should you also let us know what test cases you've conducted with the code? i.e.: 

1. Includes work when no excludes are defined.
2. Includes and excludes work when there's no overlap.
3. Includes work when they're a subset of excludes.
4. etc.

**Comment by kingthorin** [comment]
@Dalimil any update?

**Comment by kingthorin** [comment]
@Dalimil any update?

**Comment by Dalimil** [comment]
So the behaviour was that the include list can overrule the exclude list, so whenever it checks if a URL is excluded in excludeUrl(uri) it would check the exclude list for any matches as before but now whenever it is about to return true (as in 'ignore this url') it would also check the include list for possible matches and might eventually return false ('don't ignore this url') from the original function if it found any.

But looking at the code now, the functionality might be a bit confusing for many users...  
I might just drop this PR because I haven't touched the codebase in a long time and there have been many changes to the project since August...

**Comment by kingthorin** [comment]
Closing per @Dalimil

**Comment by psiinon** [comment]
For info just added the 'invalid' tag so that it doesnt get picked up for the release notes ;)

**Pull request closed without merging**